### PR TITLE
[READY] Update github webhooks to the TG's latest implementation

### PR DESCRIPTION
**Pull request by Couls**
**What does this PR do:**
Updates the github webhook code to TGs latest which allows maintainers and heads to make the bot capable of adding and removing labels from PRs automatically(PRs with balance in CL will automatically get balance label) instead of manually doing it themselves. The code also marks code ready for review and marks code with merge conflict and also has an experimental and optional PR balance system where you get points for fixing things and can spend points from your PR balance on features and rebalancing

Secrets.php needs to be configured to use the settings we already use for the old webhook system(apikey and secrets token) and the PR balance and discord integration is already off by default in game announcements are also off by default

TESTED AND WORKING!
image
image
image

https://github.com/Couls/Paradise/pulls - Go ahead and make a PR with a CL and it'll automatically apply tags! (If my webhook server is up)

**Changelog:**
:cl:
tweak: our webhooks now use TG's latest implementation
/:cl:

### CODE REVIEW

**Comment by Kyep** [comment]
Thoughts:
- This needs a justification / explanation of features, so we aren't left reading it all trying to work out what it all does.
- This needs config files split into actual config files and example config files, so that actual config files aren't in github (since they contain passwords/etc) but there's still a base to use for them

**Comment by Couls** [comment]
Secret.php are the config files that can be moved so the passwords and the like are hidden, The biggest new feature is that it automatically searches for keywords and adds labels according to them, it also adds a label if there’s a merge conflict
the files already are pretty well commented especially the variables inside secret.php so it already works as a good example config template though I could rename it to example.php instead

**Comment by AffectedArc07** [comment]
See #10430 

Maints didnt like this being merged without ***thorough*** testing. So unless you have heavily tested with your own repo and PHP server, I would be careful

**Comment by Kyep** [comment]
Since you've looked at the API: is it possible to auto-tag PRs as "Conflicted" when they have a merge conflict, and un-tag them once the conflict is resolved?

**Comment by Couls** [comment]
Yep, this is part of the update

File: tools/WebhookProcessor/github_webhook_processor.php [hunk 1]
```diff
@@ -0,0 +1,1023 @@
...
+		if(strpos(strtolower($title), 'revert') !== FALSE)
+			$tags[] = 'Revert';
+		if(strpos(strtolower($title), 'removes') !== FALSE)
+			$tags[] = 'Removal';
+	}
+
+	$remove = array('Test Merge Candidate');
+
+	$mergeable = $payload['pull_request']['mergeable'];
+	if($mergeable === TRUE)	//only look for the false value
```
Comment (Couls) [inline tools/WebhookProcessor/github_webhook_processor.php]: Here is the check for merge ability

Comment (Kyep) [reply tools/WebhookProcessor/github_webhook_processor.php]: Have you actually tested whether this checks for merge conflict or not?

**Comment by AffectedArc07** [comment]
I’m using this hook in production, and I can tell you it does indeed check

Sent from my iPhone

On 15 Aug 2019, at 08:41, Kyep <notifications@github.com> wrote:

﻿

@Kyep commented on this pull request.

________________________________

In tools/WebhookProcessor/github_webhook_processor.php<https://github.com/ParadiseSS13/Paradise/pull/11781#discussion_r314203861>:

> +     if($opened) {   //you only have one shot on these ones so as to not annoy maintainers
+               $tags = checkchangelog($payload, false);
+
+               if(strpos(strtolower($title), 'refactor') !== FALSE)
+                       $tags[] = 'Refactor';
+
+               if(strpos(strtolower($title), 'revert') !== FALSE)
+                       $tags[] = 'Revert';
+               if(strpos(strtolower($title), 'removes') !== FALSE)
+                       $tags[] = 'Removal';
+       }
+
+       $remove = array('Test Merge Candidate');
+
+       $mergeable = $payload['pull_request']['mergeable'];
+       if($mergeable === TRUE) //only look for the false value


Have you actually tested whether this checks for merge conflict or not?

**Comment by AffectedArc07** [comment]
Good fucking lord the github email reply system and iOS do NOT play nicely with each other

File: tools/WebhookProcessor/github_webhook_processor.php (hunk 1 above)
Comment (Couls) [reply tools/WebhookProcessor/github_webhook_processor.php]: That’s not one of the things I tested but DTX is using the same version and it works for him, if you require me to test it I can do so when I have time though

Comment (AffectedArc07) [reply tools/WebhookProcessor/github_webhook_processor.php]: image
Image of check working

Comment (Shadow-Quill) [reply tools/WebhookProcessor/github_webhook_processor.php]: Were you surprised?

Comment (AffectedArc07) [reply tools/WebhookProcessor/github_webhook_processor.php]: ?

**Pull request closed without merging**